### PR TITLE
DOTNET-4509 Integration tests for metrics

### PR DESCRIPTION
**Pull request by vuqtran88**

### CODE REVIEW

File: src/IntegrationTests/IntegrationTests/OpenTelemetrySmokeTest.cs [hunk 1]
```diff
@@ -88,5 +89,31 @@ public async void Test()
...
 
         }
+
+        [Fact]
+        public async void MetricTest()
+        {
+            using var httpClient = new HttpClient();
+
+            // SELECT * FROM Metric WHERE metricName = 'WeatherForecast/Get' SINCE 2 minutes ago
+            var insightQuery = "SELECT%20%2A%20FROM%20Metric%20WHERE%20metricName%20%3D%20%27WeatherForecast%2FGet%27%20SINCE%202%20minutes%20ago";
```
Comment (jifeingo) [inline src/IntegrationTests/IntegrationTests/OpenTelemetrySmokeTest.cs]: Why not build the string and then URLEncode it?

File: src/NewRelic.Telemetry/Metrics/Metric.cs [hunk 2]
```diff
@@ -40,24 +40,28 @@ public abstract class Metric
         /// <summary>
         /// TODO
         /// </summary>
+        [DataMember(Name = "name")]
```
Comment (jifeingo) [inline src/NewRelic.Telemetry/Metrics/Metric.cs]: This should not be needed

File: src/NewRelic.Telemetry/Metrics/Metric.cs [hunk 3]
```diff
@@ -40,24 +40,28 @@ public abstract class Metric
...
         public string Name { get; internal set; }
 
         /// <summary>
         /// TODO
         /// </summary>
+        [DataMember(Name = "type")]
```
Comment (jifeingo) [inline src/NewRelic.Telemetry/Metrics/Metric.cs]: This shouldn't be needed

File: src/NewRelic.Telemetry/Metrics/Metric.cs [hunk 4]
```diff
@@ -40,24 +40,28 @@ public abstract class Metric
...
         public abstract string Type { get; }
 
         /// <summary>
         /// TODO
         /// </summary>
-        /// 
+        ///
+        [DataMember(Name = "timestamp")]
```
Comment (jifeingo) [inline src/NewRelic.Telemetry/Metrics/Metric.cs]: This shouldn't be needed

File: src/IntegrationTests/Applications/SampleAspNetCoreApp/Controllers/WeatherForecastController.cs [hunk 5]
```diff
@@ -19,6 +28,21 @@ public class WeatherForecastController : ControllerBase
...
+            .WithIntervalMs(10)
+            .WithAttribute("testAttributeKey1", "testAttributeValue1");
+            
+            var metric = metricBuilder.Build();
+
+            var metricBatch = MetricBatchBuilder.Create()
+            .WithMetric(metric)
+            .Build();
+
+            var dataSender = new MetricDataSender(_config);
```
Comment (jifeingo) [inline src/IntegrationTests/Applications/SampleAspNetCoreApp/Controllers/WeatherForecastController.cs]: How do you feel about using Dependency Injection to inject this into the controller but resolve it in the container as a singleton.?

File: src/IntegrationTests/Applications/SampleAspNetCoreApp/Controllers/WeatherForecastController.cs [hunk 6]
```diff
@@ -3,14 +3,23 @@
...
 
 namespace SampleAspNetCoreApp.Controllers
 {
     [ApiController]
     [Route("[controller]")]
     public class WeatherForecastController : ControllerBase
     {
+        private readonly IConfiguration _config;
+
+        public WeatherForecastController(IConfiguration config)
```
Comment (jifeingo) [inline src/IntegrationTests/Applications/SampleAspNetCoreApp/Controllers/WeatherForecastController.cs]: Let's use dependency injection to build the metric data sender and send it in.
Can we do the same for the Opentelemetry stuff if it is not already like this.

File: src/IntegrationTests/Applications/SampleAspNetCoreApp/CountMetricGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            .WithValue(1)
+            .WithIntervalMs(10);
+
+            var metric = metricBuilder.Build();
+
+            var metricBatch = MetricBatchBuilder.Create()
+            .WithMetric(metric)
+            .Build();
+
+            var dataSender = new MetricDataSender(_config);
```
Comment (jifeingo) [inline src/IntegrationTests/Applications/SampleAspNetCoreApp/CountMetricGenerator.cs]: This should be dependency injected as well.  It should be a singleton

File: src/IntegrationTests/Applications/SampleAspNetCoreApp/Startup.cs [hunk 8]
```diff
@@ -47,6 +47,8 @@ public void ConfigureServices(IServiceCollection services)
                 var dependenciesCollector = new DependenciesCollector(new HttpClientCollectorOptions(), tracerFactory);
                 var aspNetCoreCollector = new AspNetCoreCollector(tracerFactory.GetTracer(null));
             });
+
```
Comment (jifeingo) [inline src/IntegrationTests/Applications/SampleAspNetCoreApp/Startup.cs]: If you were going to dependency inject the data sender, you would declare it here.
then the CountMetricGenerator would take the datasender in its constructor and the dependency injection container is responsible for resolving it.

File: src/IntegrationTests/Applications/SampleAspNetCoreApp/CountMetricGenerator.cs (hunk 7 above)
Comment (vuqtran88) [reply src/IntegrationTests/Applications/SampleAspNetCoreApp/CountMetricGenerator.cs]: 👍

File: src/IntegrationTests/Applications/SampleAspNetCoreApp/CountMetricGenerator.cs [hunk 9]
```diff
@@ -7,11 +7,11 @@ namespace SampleAspNetCoreApp
 {
     public class CountMetricGenerator
     {
-        private IConfiguration _config;
+        private MetricDataSender _metricDataSender;
```
Comment (jifeingo) [inline src/IntegrationTests/Applications/SampleAspNetCoreApp/CountMetricGenerator.cs]: readonly

**Pull request merged**